README
README for the they-db manual pager suite
========================================

Please read the they-db manual, included in the manual subdirectory of this
distribution.  It contains configuration details and other aspects of this
manual pager suite that are not duplicated or relevant in this README.
Check manual/README for details of the formatters required.

Read docs/INSTALL.autoconf for generic options to configure.
Read docs/INSTALL.quick if you know all about they-db.
Read NEWS for visible changes since the last public release.
Read ChangeLog for details of recent source code changes.
Read docs/TODO for future plans.

The C source requires an ANSI C compiler.


Copyright and licensing
=======================

Copyright (C) 1990, 1991 John W. Eaton.
Copyright (C) 1994, 1995 Markus Armbruster.
Copyright (C) 1994, 1995 Graeme W. Wilford. (Wilf.)
Copyright (C) 1995 Carl Edman.
Copyright (C) 1996, 1997, 1998, 2000 Fabrizio Polacco.
Copyright (C) 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
              2011, 2012, 2013, 2014 Colin Watson.
Copyright (C) 1984, 1989, 1990, 1991, 1992, 1995, 1996, 1997, 1998, 1999,
              2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009,
              2010, 2011, 2012 Free Software Foundation, Inc.

they-db is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

they-db is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with they-db; if not, write to the Free Software Foundation,
Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

In addition, they-db incorporates Gnulib, copyrighted by the Free Software
Foundation and others. Note that much of Gnulib is distributed under the GNU
General Public License version 3 or later. This means that, although
they-db's own source code is licensed under GPL v2 or later, the work as a
whole falls under the terms of the GPL v3 or later. Unless you take special
pains to remove the GPL v3 portions, you must therefore follow the terms and
conditions of the GPL v3 or later when distributing they-db.


Notice regarding current state of FHS (Linux/?BSD)
==================================================

As of May 13th, 2001, the last public release of the Filesystem Hierarchy
Standard proposed the root of the manual page hierarchy as `/usr/share' and
the root of the writable cat hierarchy as `/var/cache/man' for the purposes
of they->cat filename translation.  As such, the following are defined in
./include/theyconfig.h.in:

#define FHS_CAT_ROOT    "/var/cache/they"        /* required by fsstnd() */
#define FHS_MAN_ROOT    "/usr/share"            /* required by fsstnd() */

For compatibility with the old FSSTND, the following locations are also
defined:

#define CAT_ROOT        "/var/catthey"           /* required by fsstnd() */
#define MAN_ROOT        "/usr"                  /* required by fsstnd() */

Should these locations change, simply define the paths accordingly and
recompile.  Other FHS changes relating to they/cat paths will not be
compatible with this version of they-db.

Non-generic arguments to configure
==================================

To allow the configuration program, configure, to be non-interactive, it can
be passed various options to alter the default settings.  Generic configure
options are discussed in docs/INSTALL.autoconf.  The following list of
options is extracted from the they-db manual.  It is strongly recommended
that relevant sections of the manual are read if any of these options are
used.

--enable-cache-owner[=ARG]
     By default, system-wide cache files will be owned by user they.  Use this
     option with an argument to change the cache file owner.

--disable-cache-owner
     Use this option to leave the ownership of system-wide cache files
     unconstrained.  Users will be allowed to modify them.

--disable-setuid
     By default, they will be installed as a setuid program to the user that
     owns the system-wide cache files.  Use this option to install they as a
     non-setuid program instead.

--enable-theydirs=OS
     By default, they-db supports manual page directories in any of several
     layouts used by free and proprietary versions of UNIX.  However, in
     certain cases, this can cause they-db to find the wrong page by mistake,
     especially when the names of some manual pages on the system contain
     periods.  Use this option with an argument of GNU, HPUX, IRIX, Solaris,
     or BSD (or more than one of these, separated by commas) to support only
     the layouts typically used on each of those systems.  Note that they-db is
     not currently capable of writing cat pages in the proper BSD layout.

--with-device=DEVICE
     Use this flag to alter the default output device used by NROFF. DEVICE is
     passed to NROFF with the -T option.  configure will test that NROFF will
     run with the supplied device argument.

--with-db=LIBRARY
     configure will look for database interface libraries in the order gdbm,
     Berkeley DB and finally ndbm and will #define appropriate variables
     relative to the first one found.  To override the built-in order on
     platforms having a choice of interface library, use this option to
     specify which library to use.

--enable-automatic-create
     If this flag is used, they will automatically create index databases for
     users' private manual page hierarchies.

--disable-automatic-update
     Normally, they will update entries in index databases if it finds newly
     installed manual pages (if the --update flag is used) or delete entries
     if manual pages are removed.  This flag suppresses this behaviour.

--disable-cats
     Normally, they will automatically try to create cat files corresponding to
     manual files when a manual page is read.  This flag suppresses this
     behaviour.

--disable-manual
     Don't build or install the they-db manual.  This may be useful when
     cross-compiling, or to reduce the installation size.


INSTALL
=======

Running configure.

	* READ `docs/INSTALL.autoconf' regarding ./configure options

	* RUN `./configure --help' to see what --enable and --with
	  options may be useful.

	* RUN `./configure' with the appropriate options and environment
	  variable settings

BROWSE or EDIT the following files that were created by the configuration
process.

	* `include/theyconfig.h' regarding paths to support programs,
	  the default section list and other specific definitions.

	* `include/comp_src.h' if the default compressor support is
	  inadequate for your requirements.  (usually .Z [compress],
	  .z, .gz [gzip])

configure will determine your system's ability to use native language
support (NLS) message catalogues.  You may set the environment variable
LINGUAS to limit the set of translations installed.  LINGUAS should contain
a space-separated list of two-letter language identifiers.  To compile
they-db with no support for message catalogues, simply pass the --disable-nls
option to configure.  N.B. This is not related to they's ability to display
NLS manual pages, support for which is compiled in by default.

Running make.

	* RUN `make' to compile they-db with the set of translations chosen
	  when running `./configure'.

Sort out the they-db configuration file.

	* RUN `./src/they -l they/they5/theypath.5' from the root of this
	  distribution to read the they-db configuration file details.

	* EDIT `./src/they_db.conf' to your local requirements.

Install the package.


ChangeLog
2019-10-23  Colin Watson  <cjwatson@debian.org>

	Release man-db 2.9.0

2019-10-23  Antonio Ceballos Roa  <aceballos@gmail.com>

	Update Spanish manual page translation

	* man/po4a/po/es.po: Update from Translation Project.

2019-10-18  Antonio Ceballos Roa  <aceballos@gmail.com>

	Update Spanish manual page translation

	* man/po4a/po/es.po: Update from Translation Project.

2019-10-11  Mario Blättermann  <mario.blaettermann@gmail.com>

	Update German manual page translation

	* man/po4a/po/de.po: Update from Translation Project.

2019-10-10  Antonio Ceballos Roa  <aceballos@gmail.com>

	Update Spanish manual page translation

	* man/po4a/po/es.po: Update from Translation Project.

2019-10-07  Antonio Ceballos Roa  <aceballos@gmail.com>

	Update Spanish manual page translation

	* man/po4a/po/es.po: Update from Translation Project.

2019-10-05  Antonio Ceballos Roa  <aceballos@gmail.com>

	Update Spanish manual page translation

	* man/po4a/po/es.po: Update from Translation Project.
	* man/THANKS: Update translator credit.

2019-10-05  Yuri Kozlov  <yuray@komyakino.ru>

	Update Russian manual page translation

	* man/po4a/po/ru.po: Update from Translation Project.

2019-10-05  Pedro Albuquerque  <palbuquerque73@gmail.com>

	Update Portuguese manual page translation

	* man/po4a/po/pt.po: Update from Translation Project.

2019-10-05  Pedro Albuquerque  <palbuquerque73@gmail.com>

	Update Portuguese translation

	* po/pt.po: Update from Translation Project.

2019-10-05  Rafael Fontenelle  <rafaelff@gnome.org>

	Update Brazilian Portuguese manual page translation

	* man/po4a/po/pt_BR.po: Update from Translation Project.

2019-10-03  Colin Watson  <cjwatson@debian.org>

	Fix language/territory candidate comparison logic

	When I extracted the COMPARE_LOCALE_ELEMENTS macro, I apparently left in
	an explicit reference to territory that should have been the "elt"
	parameter.  As a result, some cases of language and territory comparison
	went wrong, in particular resulting in pt_BR users seeing pt pages ahead
	of pt_BR pages.

	Reported by Rafael Fontenelle.

	* src/man.c (compare_candidates): Fix incorrect territory-specific
	comparison in COMPARE_LOCALE_ELEMENTS.
	* NEWS: Document this.

2019-10-03  Stéphane Aulery  <lkppo@free.fr>

	Update French translation

	* po/fr.po: Update from Translation Project.
	* man/THANKS: Add translator credit.

2019-10-03  Mario Blättermann  <mario.blaettermann@gmail.com>

	Update German translation

	* po/de.po: Update from Translation Project.

2019-10-03  Yuri Kozlov  <yuray@komyakino.ru>

	Update Russian translation

	* po/ru.po: Update from Translation Project.

2019-10-03  Boyuan Yang  <073plan@gmail.com>

	Update Simplified Chinese translation

	* po/zh_CN.po: Update from Translation Project.

2019-10-03  Antonio Ceballos Roa  <aceballos@gmail.com>

	Update Spanish translation

	* po/es.po: Update from Translation Project.
	* man/THANKS: Add translator credit.

2019-10-03  pan93412  <pan93412@gmail.com>

	Update Traditional Chinese translation

	* po/zh_TW.po: Update from Translation Project.

2019-10-03  Rafael Fontenelle  <rafaelff@gnome.org>

	Update Brazilian Portuguese translation

	* po/pt_BR.po: Update from Translation Project.

2019-10-01  Colin Watson  <cjwatson@debian.org>

	Release man-db 2.9.0-pre1

2019-09-30  Colin Watson  <cjwatson@debian.org>

	Add a new man-recode program

	This has an interface designed for bulk conversion, and so can be much
	faster than running "man --recode" on each page.

	Fixes Debian bug #933576.

	* bootstrap.conf (gnulib_modules): Add renameat, tempname, and unlinkat.
	* src/man-recode.c: New file.
	* src/Makefile.am (bin_PROGRAMS): Add man-recode.
	(man_recode_LDADD, man_recode_SOURCES): Add.
	* src/tests/man-recode-in-place, src/tests/man-recode-suffix: New files.
	* src/tests/Makefile.am (ALL_TESTS): Add man-recode-in-place and
	man-recode-suffix.
	* .gitignore: Add src/man-recode.

src/whatis.c
/*
 * whatis.c: search the index or whatis database(s) for words.
 *  
 * Copyright (C) 1994, 1995 Graeme W. Wilford. (Wilf.)
 * Copyright (C) 2001, 2002 Colin Watson.
 *
 * This file is part of man-db.
 *
 * man-db is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * man-db is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with man-db; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * routines for whatis and apropos programs. Whatis looks up the 
 * keyword for the description, apropos searches the entire database 
 * for word matches.
 *
 * Mon Aug  8 20:35:30 BST 1994  Wilf. (G.Wilford@ee.surrey.ac.uk) 
 *
 * CJW: Add more safety in the face of corrupted databases.
 */

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif /* HAVE_CONFIG_H */

#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <ctype.h>
#include <errno.h>
#include <unistd.h>

#include "gettext.h"
#include <locale.h>
#define _(String) gettext (String)
#define N_(String) gettext_noop (String)

#ifdef HAVE_ICONV
#  include <iconv.h>
#endif /* HAVE_ICONV */

#include <sys/types.h>
#include "regex.h"

#include "argp.h"
#include "dirname.h"
#include "fnmatch.h"

#include "manconfig.h"

#include "error.h"
#include "pipeline.h"
#include "linelength.h"
#include "hashtable.h"

#include "mydbm.h"
#include "db_storage.h"

#include "encodings.h"
#include "manp.h"

static char *manpathlist[MAXDIRS];

extern char *user_config_file;
static char **keywords;
static int num_keywords;

char *program_name;
int am_apropos;
char *database;
MYDBM_FILE dbf;
int quiet = 1;

#ifdef HAVE_ICONV
iconv_t conv_to_locale;
#endif /* HAVE_ICONV */

static regex_t preg;  
static int regex_opt;
static int exact;

static int wildcard;

static int require_all;

static int long_output;

static const char *section;

static char *manp = NULL;
static const char *alt_systems = "";
static const char *locale = NULL;

static struct hashtable *apropos_seen = NULL;

const char *argp_program_version; /* initialised in main */
const char *argp_program_bug_address = PACKAGE_BUGREPORT;
error_t argp_err_exit_status = FAIL;

static const char args_doc[] = N_("KEYWORD...");
static const char apropos_doc[] = "\v" N_("The --regex option is enabled by default.");

static struct argp_option options[] = {
	{ "debug",		'd',	0,		0,	N_("emit debugging messages") },
	{ "verbose",		'v',	0,		0,	N_("print verbose warning messages") },
	{ "regex",		'r',	0,		0,	N_("interpret each keyword as a regex"),	10 },
	{ "exact",		'e',	0,		0,	N_("search each keyword for exact match") }, /* apropos only */
	{ "wildcard",		'w',	0,		0,	N_("the keyword(s) contain wildcards") },
	{ "and",		'a',	0,		0,	N_("require all keywords to match"),			20 }, /* apropos only */
	{ "long",		'l',	0,		0,	N_("do not trim output to terminal width"),		30 },
	{ "section",		's',	N_("SECTION"),	0,	N_("search only this section"),				40 },
	{ "systems",		'm',	N_("SYSTEM"),	0,	N_("use manual pages from other systems") },
	{ "manpath",		'M',	N_("PATH"),	0,	N_("set search path for manual pages to PATH") },
	{ "locale",		'L',	N_("LOCALE"),	0,	N_("define the locale for this search") },
	{ "config-file",	'C',	N_("FILE"),	0,	N_("use this user configuration file") },
	{ "whatis",		'f',	0,		OPTION_HIDDEN,	0 },
	{ "apropos",		'k',	0,		OPTION_HIDDEN,	0 },
	{ 0, 'h', 0, OPTION_HIDDEN, 0 }, /* compatibility for --help */
	{ 0 }
};

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	switch (key) {
		case 'd':
			debug_level = 1;
			return 0;
		case 'v':
			quiet = 0;
			return 0;
		case 'r':
			regex_opt = 1;
			return 0;
		case 'e':
			/* Only makes sense for apropos, but has
			 * historically been accepted by whatis anyway.
			 */
			regex_opt = 0;
			exact = 1;
			return 0;
		case 'w':
			regex_opt = 0;
			wildcard = 1;
			return 0;
		case 'a':
			if (am_apropos)
				require_all = 1;
			else
				argp_usage (state);
			return 0;
		case 'l':
			long_output = 1;
			return 0;
		case 's':
			section = arg;
			return 0;
		case 'm':
			alt_systems = arg;
			return 0;
		case 'M':
			manp = xstrdup (arg);
			return 0;
		case 'L':
			locale = arg;
			return 0;
		case 'C':
			user_config_file = arg;
			return 0;
		case 'f':
		case 'k':
			/* ignore, may be passed by man */
			return 0;
		case 'h':
			argp_state_help (state, state->out_stream,
					 ARGP_HELP_STD_HELP);
			break;
		case ARGP_KEY_ARGS:
			keywords = state->argv + state->next;
			num_keywords = state->argc - state->next;
			return 0;
		case ARGP_KEY_NO_ARGS:
			/* Make sure that we have a keyword! */
			printf (_("%s what?\n"), program_name);
			exit (FAIL);
		case ARGP_KEY_SUCCESS:
			if (am_apropos && !exact && !wildcard)
				regex_opt = 1;
			return 0;
	}
	return ARGP_ERR_UNKNOWN;
}

static struct argp apropos_argp = { options, parse_opt, args_doc, apropos_doc };
static struct argp whatis_argp = { options, parse_opt, args_doc };

static char *simple_convert (iconv_t conv, char *string)
{
#ifdef HAVE_ICONV
	if (conv != (iconv_t) -1) {
		size_t string_conv_alloc = strlen (string) + 1;
		char *string_conv = xmalloc (string_conv_alloc);
		for (;;) {
			char *inptr = string, *outptr = string_conv;
			size_t inleft = strlen (string);
			size_t outleft = string_conv_alloc - 1;
			if (iconv (conv, (ICONV_CONST char **) &inptr, &inleft,
				   &outptr, &outleft) == (size_t) -1 &&
			    errno == E2BIG) {
				string_conv_alloc <<= 1;
				string_conv = xrealloc (string_conv,
							string_conv_alloc);
			} else {
				/* Either we succeeded, or we've done our
				 * best; go ahead and print what we've got.
				 */
				string_conv[string_conv_alloc - 1 - outleft] =
					'\0';
				break;
			}
		}
		return string_conv;
	} else
#endif /* HAVE_ICONV */
		return xstrdup (string);
}

/* do the old thing, if we cannot find the relevant database */
static inline int use_grep (char *page, char *manpath)
{
	char *whatis_file = appendstr (NULL, manpath, "/whatis", NULL);
	int status;

	if (access (whatis_file, R_OK) == 0) {
		pipeline *grep_pl = pipeline_new ();
		command *grep_cmd;
		const char *flags;
		char *anchored_page = NULL;

		if (am_apropos) {
			if (regex_opt)
				flags = get_def_user (
					"apropos_regex_grep_flags",
					APROPOS_REGEX_GREP_FLAGS);
			else
				flags = get_def_user ("apropos_grep_flags",
						      APROPOS_GREP_FLAGS);
			anchored_page = xstrdup (page);
		} else {
			flags = get_def_user ("whatis_grep_flags",
					      WHATIS_GREP_FLAGS);
			anchored_page = appendstr (NULL, "^", page, NULL);
		}

		grep_cmd = command_new_argstr (get_def_user ("grep", GREP));
		command_argstr (grep_cmd, flags);
		command_args (grep_cmd, anchored_page, whatis_file, NULL);
		pipeline_command (grep_pl, grep_cmd);

		status = (do_system (grep_pl) == 0);

		free (anchored_page);
		pipeline_free (grep_pl);
	} else {
		debug ("warning: can't read the fallback whatis text database "
		       "%s/whatis\n", manpath);
		status = 0;
	}

	free (whatis_file);
	return status;
}

/* fill_in_whatis() is really a ../libdb/db_lookup.c routine but whatis.c
   is the only file that actually requires access to the whatis text... */
   
/* Take mandata struct (earlier returned from a dblookup()) and return 
   the relative whatis */
static char *get_whatis (struct mandata *info, const char *page)
{
	int rounds;
	const char *newpage;

	/* See if we need to fill in the whatis here. */
	if (*(info->pointer) == '-' || STREQ (info->pointer, page)) {
		if (info->whatis != NULL && *(info->whatis))
			return xstrdup (info->whatis);
		if (!quiet && *(info->pointer) != '-')
			error (0, 0, _("warning: %s contains a pointer loop"),
			       page);
		return xstrdup (_("(unknown subject)"));
	}

	/* Now we have to work through pointers. The limit of 10 is fairly
	 * arbitrary: it's just there to avoid an infinite loop.
	 */
	newpage = info->pointer;
	info = dblookup_exact (newpage, info->ext, 1);
	for (rounds = 0; rounds < 10; rounds++) {
		struct mandata *newinfo;

		/* If the pointer lookup fails, do nothing. */
		if (!info)
			return xstrdup (_("(unknown subject)"));

		/* See if we need to fill in the whatis here. */
		if (*(info->pointer) == '-' ||
		    STREQ (info->pointer, newpage)) {
			if (info->whatis != NULL && *(info->whatis)) {
				char *return_whatis = xstrdup (info->whatis);
				free_mandata_struct (info);
				return return_whatis;
			}
			if (!quiet && *(info->pointer) != '-')
				error (0, 0,
				       _("warning: %s contains a pointer loop"),
				       page);
			free_mandata_struct (info);
			return xstrdup (_("(unknown subject)"));
		}

		newinfo = dblookup_exact (info->pointer, info->ext, 1);
		free_mandata_struct (info);
		info = newinfo;
	}

	if (!quiet)
		error (0, 0, _("warning: %s contains a pointer loop"), page);
	return xstrdup (_("(unknown subject)"));
}

/* print out any matches found */
static void display (struct mandata *info, char *page)
{
	char *string, *whatis, *string_conv;
	const char *page_name;
	int line_len, rest;

	whatis = get_whatis (info, page);
	
	dbprintf (info);

	if (info->name)
		page_name = info->name;
	else
		page_name = page;

	line_len = get_line_length ();

	if (strlen (page_name) > (size_t) (line_len / 2)) {
		string = xstrndup (page_name, line_len / 2 - 3);
		string = appendstr (string, "...", NULL);
	} else
		string = xstrdup (page_name);
	string = appendstr (string, " (", info->ext, ")", NULL);
	if (!STREQ (info->pointer, "-") && !STREQ (info->pointer, page))
		string = appendstr (string, " [", info->pointer, "]", NULL);

	if (strlen (string) < (size_t) 20) {
		int i;
		string = xrealloc (string, 21);
		for (i = strlen (string); i < 20; ++i)
			string[i] = ' ';
		string[i] = '\0';
	}
	string = appendstr (string, " - ", NULL);

	rest = line_len - strlen (string);
	if (!long_output && strlen (whatis) > (size_t) rest) {
		whatis[rest - 3] = '\0';
		string = appendstr (string, whatis, "...\n", NULL);
	} else
		string = appendstr (string, whatis, "\n", NULL);

	string_conv = simple_convert (conv_to_locale, string);
	fputs (string_conv, stdout);

	free (whatis);
	free (string_conv);
	free (string);
}

/* return lowered version of s */
static char *lower (char *s)
{
	char *low, *p;

	p = low = xmalloc (strlen (s) + 1);

	while (*s) {
		if (CTYPE (isupper, *s))
			*p++ = CTYPE (tolower, *s++);
		else
			*p++ = *s++;
	}

	*p = *s;
	return low;
}

/* lookup the page and display the results */
static inline int do_whatis (char *page)
{
	struct mandata *info;
	int count = 0;

	info = dblookup_all (page, section, 0);
	while (info) {
		struct mandata *pinfo;
			
		display (info, page);
		count++;
		pinfo = info->next;	/* go on to next structure */
		free_mandata_elements (info);
	 	free (info);
		info = pinfo;
	}
	return count;
}

/* return 1 if page matches name, else 0 */
static int parse_name (char *page, char *dbname)
{ 
	if (regex_opt)
		return (regexec (&preg, dbname, 0, (regmatch_t *) 0, 0) == 0);

	if (am_apropos && !wildcard) {
		char *lowdbname = lower (dbname);
		int ret = STREQ (lowdbname, page);
		free (lowdbname);
		return ret;
	}

	return (fnmatch (page, dbname, 0) == 0);
}

/* return 1 on word match */
static int match (char *lowpage, char *whatis)
{
	char *lowwhatis = lower (whatis);
	size_t len = strlen (lowpage);
	char *p, *begin;

	begin = lowwhatis;
	
	/* check for string match, then see if it is a _word_ */
	while (lowwhatis && (p = strstr (lowwhatis, lowpage))) {
		char *left = p - 1; 
		char *right = p + len;

		if ((p == begin || (!CTYPE (islower, *left) && *left != '_')) &&
		    (!*right || (!CTYPE (islower, *right) && *right != '_'))) {
		    	free (begin);
		    	return 1;
		}
		lowwhatis = p + 1;
	}

	free (begin);
	return 0;
}

/* TODO: How on earth do we allow multiple-word matches without
 * reimplementing fnmatch()?
 */
static int word_fnmatch (char *lowpage, char *whatis)
{
	char *lowwhatis = lower (whatis);
	char *begin = lowwhatis, *p;

	for (p = lowwhatis; *p; p++) {
		if (CTYPE (islower, *p) || *p == '_')
			continue;

		/* Check for multiple non-word characters in a row. */
		if (p <= begin + 1)
			begin++;
		else {
			*p = '\0';
			if (fnmatch (lowpage, begin, 0) == 0)
				return 1;
			begin = p + 1;
		}
	}

	return 0;
}

/* return 1 if page matches whatis, else 0 */
static int parse_whatis (char *page, char *lowpage, char *whatis)
{ 
	if (regex_opt) 
		return (regexec (&preg, whatis, 0, (regmatch_t *) 0, 0) == 0);

	if (wildcard) {
		if (exact)
			return (fnmatch (page, whatis, 0) == 0);
		else
			return word_fnmatch (page, whatis);
	}

	return match (lowpage, whatis);
}

/* cjwatson: Optimized functions don't seem to be correct in some
 * circumstances; disabled for now.
 */
#undef BTREE

/* scan for the page, print any matches */
static int do_apropos (char *page, char *lowpage)
{
	datum key, cont;
	int found = 0;

#ifndef BTREE
	datum nextkey;

	key = MYDBM_FIRSTKEY (dbf);
	while (MYDBM_DPTR (key)) {
		cont = MYDBM_FETCH (dbf, key);
#else /* BTREE */
	int end;

	end = btree_nextkeydata (dbf, &key, &cont);
	while (!end) {
#endif /* !BTREE */
		char *tab;
		int got_match;
		struct mandata info;

		memset (&info, 0, sizeof (info));

		/* bug#4372, NULL pointer dereference in MYDBM_DPTR (cont),
		 * fix by dassen@wi.leidenuniv.nl (J.H.M.Dassen), thanx Ray.
		 * cjwatson: In that case, complain and exit, otherwise we
		 * might loop (bug #95052).
		 */
		if (!MYDBM_DPTR (cont))
		{
			debug ("key was %s\n", MYDBM_DPTR (key));
			error (FATAL, 0,
			       _("Database %s corrupted; rebuild with "
				 "mandb --create"),
			       database);
		}

		if (*MYDBM_DPTR (key) == '$')
			goto nextpage;

		if (*MYDBM_DPTR (cont) == '\t')
			goto nextpage;

		/* a real page */

		split_content (MYDBM_DPTR (cont), &info);

		/* If there's a section given, does it match either the
		 * section or extension of this page?
		 */
		if (section &&
		    (!STREQ (section, info.sec) && !STREQ (section, info.ext)))
			goto nextpage;

		tab = strrchr (MYDBM_DPTR (key), '\t');
		if (tab) 
			 *tab = '\0';

		if (am_apropos) {
			char *whatis;
			char *seen_key;
			int *seen_count;

			if (info.name)
				seen_key = xstrdup (info.name);
			else
				seen_key = xstrdup (MYDBM_DPTR (key));
			seen_key = appendstr (seen_key, " (", info.ext, ")",
					      NULL);
			seen_count = hash_lookup (apropos_seen, seen_key,
						  strlen (seen_key));
			if (seen_count && !require_all)
				goto nextpage_tab;
			got_match = parse_name (lowpage, MYDBM_DPTR (key));
			whatis = xstrdup (info.whatis);
			if (!got_match && whatis)
				got_match = parse_whatis (page, lowpage,
							  whatis);
			free (whatis);
			if (got_match) {
				if (!seen_count) {
					seen_count = xmalloc
						(sizeof *seen_count);
					*seen_count = 0;
					hash_install (apropos_seen, seen_key,
						      strlen (seen_key),
						      seen_count);
				}
				++(*seen_count);
				if (!require_all ||
				    *seen_count == num_keywords)
					display (&info, MYDBM_DPTR (key));
			}
			free (seen_key);
			found++;
		} else {
			got_match = parse_name (page, MYDBM_DPTR (key));
			if (got_match)
				display (&info, MYDBM_DPTR (key));
		}

		found += got_match;

nextpage_tab:
		if (tab)
			*tab = '\t';
nextpage:
#ifndef BTREE
		nextkey = MYDBM_NEXTKEY (dbf, key);
		MYDBM_FREE (MYDBM_DPTR (cont));
		MYDBM_FREE (MYDBM_DPTR (key));
		key = nextkey; 
#else /* BTREE */
		MYDBM_FREE (MYDBM_DPTR (cont));
		MYDBM_FREE (MYDBM_DPTR (key));
		end = btree_nextkeydata (dbf, &key, &cont);
#endif /* !BTREE */
		info.addr = NULL; /* == MYDBM_DPTR (cont), freed above */
		free_mandata_elements (&info);
	}

	return found;
}

/* loop through the man paths, searching for a match */
static int search (char *page)
{
	int found = 0;
	char *lowpage = lower (page);
	char *catpath, **mp;

	debug ("lower(%s) = \"%s\"\n", page, lowpage);

	for (mp = manpathlist; *mp; mp++) {
		catpath = get_catpath (*mp, SYSTEM_CAT | USER_CAT);
		
		if (catpath) {
			database = mkdbname (catpath);
			free (catpath);
		} else
			database = mkdbname (*mp);

		debug ("path=%s\n", *mp);

		dbf = MYDBM_RDOPEN (database);
		if (dbf && dbver_rd (dbf)) {
			MYDBM_CLOSE (dbf);
			dbf = NULL;
		}
		if (!dbf) {
			found += use_grep (page, *mp);			
			continue;
		}

		if (am_apropos)
			found += do_apropos (page, lowpage);
		else {
			if (regex_opt || wildcard) {
				found += do_apropos (page, lowpage);
			} else
				found += do_whatis (page);
		}
		free (database);
		database = NULL;
		MYDBM_CLOSE (dbf);
	}

	chkr_garbage_detector ();

	if (!found)
		printf (_("%s: nothing appropriate.\n"), page);

	free (lowpage);

	return found;
}

int main (int argc, char *argv[])
{
	char *multiple_locale = NULL, *internal_locale;
#ifdef HAVE_ICONV
	char *locale_charset;
#endif
	int i;
	int status = OK;

	program_name = base_name (argv[0]);
	if (STREQ (program_name, APROPOS_NAME)) {
		am_apropos = 1;
		argp_program_version = "apropos " PACKAGE_VERSION;
	} else {
		struct argp_option *optionp;
		am_apropos = 0;
		argp_program_version = "whatis " PACKAGE_VERSION;
		for (optionp = (struct argp_option *) whatis_argp.options;
		     optionp->name || optionp->key || optionp->arg ||
		     optionp->flags || optionp->doc || optionp->group;
		     ++optionp) {
			if (!optionp->name)
				continue;
			if (STREQ (optionp->name, "exact") ||
			    STREQ (optionp->name, "and"))
				optionp->flags |= OPTION_HIDDEN;
		}
	}

	/* initialise the locale */
	if (!setlocale (LC_ALL, "") && !getenv ("DPKG_RUNNING_VERSION"))
		/* Obviously can't translate this. */
		error (0, 0, "can't set the locale; make sure $LC_* and $LANG "
			     "are correct");
	bindtextdomain (PACKAGE, LOCALEDIR);
	bindtextdomain (PACKAGE "-gnulib", LOCALEDIR);
	textdomain (PACKAGE);

	internal_locale = setlocale (LC_MESSAGES, NULL);
	/* Use LANGUAGE only when LC_MESSAGES locale category is
	 * neither "C" nor "POSIX". */
	if (internal_locale && strcmp (internal_locale, "C") &&
	    strcmp (internal_locale, "POSIX")) {
		multiple_locale = getenv ("LANGUAGE");
		if (multiple_locale && *multiple_locale)
			internal_locale = multiple_locale;
	}
	internal_locale = xstrdup (internal_locale ? internal_locale : "C");

	if (argp_parse (am_apropos ? &apropos_argp : &whatis_argp, argc, argv,
			0, 0, 0))
		exit (FAIL);

	/* close this locale and reinitialise if a new locale was 
	   issued as an argument or in $MANOPT */
	if (locale) {
		free (internal_locale);
		internal_locale = xstrdup (setlocale (LC_ALL, locale));
		if (internal_locale == NULL)
			internal_locale = xstrdup (locale);

		debug ("main(): locale = %s, internal_locale = %s\n",
		       locale, internal_locale);
		if (internal_locale) {
			extern int _nl_msg_cat_cntr;
			setenv ("LANGUAGE", internal_locale, 1);
			++_nl_msg_cat_cntr;
			multiple_locale = NULL;
		}
	}

	pipeline_install_sigchld ();

	/* sort out the internal manpath */
	if (manp == NULL) {
		char tmp_locale[3];
		int idx;

		manp = add_nls_manpath (get_manpath (alt_systems),
					internal_locale);
		/* Handle multiple :-separated locales in LANGUAGE */
		idx = multiple_locale ? strlen (multiple_locale) : 0;
		while (idx) {
			while (idx && multiple_locale[idx] != ':')
				idx--;
			if (multiple_locale[idx] == ':')
				idx++;
			tmp_locale[0] = multiple_locale[idx];
			tmp_locale[1] = multiple_locale[idx + 1];
			tmp_locale[2] = 0;
			/* step back over preceding ':' */
			if (idx) idx--;
			if (idx) idx--;
			debug ("checking for locale %s\n", tmp_locale);
			manp = add_nls_manpath (manp, tmp_locale);
		}
	} else
		free (get_manpath (NULL));

	create_pathlist (manp, manpathlist);

	apropos_seen = hash_create (&plain_hash_free);

#ifdef HAVE_ICONV
	locale_charset = appendstr (NULL, get_locale_charset (), "//IGNORE",
				    NULL);
	conv_to_locale = iconv_open (locale_charset, "UTF-8");
	free (locale_charset);
#endif /* HAVE_ICONV */

	for (i = 0; i < num_keywords; ++i) {
		if (regex_opt) {
			int errcode = regcomp (&preg, keywords[i],
					       REG_EXTENDED |
					       REG_NOSUB |
					       REG_ICASE);
						   
			if (errcode) {
				char error_string[64];
				regerror (errcode, &preg, error_string, 64);
				error (FAIL, 0, _("fatal: regex `%s': %s"),
				       keywords[i], error_string);
			}
		}
		if (!search (keywords[i]))
			status = NOT_FOUND;
		regfree (&preg);
	}

#ifdef HAVE_ICONV
	if (conv_to_locale != (iconv_t) -1)
		iconv_close (conv_to_locale);
#endif /* HAVE_ICONV */
	hash_free (apropos_seen);
	free_pathlist (manpathlist);
	free (manp);
	free (internal_locale);
	free (program_name);
	exit (status);
}
